### DISCUSSION TITLE
Add SEA support to pkg

### DISCUSSION THREAD

**Discussion by robertsLando** [nodejs/single-executable#78] [Ideas]
Hi everyone, I'm a maintainer of [pkg](https://github.com/vercel/pkg), I would like to know if someone could help/guide me adding SEA support to it. I have read docs about SEA but I'm not sure about how it could work together with pkg.

Actually what pkg does is (more info [here](https://github.com/vercel/pkg/wiki/Developers)):

- traverses the source code of an application and its dependencies looking for things like require()
- converts all the js files in v8 snapshots
- downloads a patched nodejs executable that will be used to create the application. Patches are needed to patch Node.js executable with a proxy around fs. This proxy adds the ability to look into something called a snapshot file system, which is where your project is stored. Patches are handled by [pkg-fetch](https://github.com/vercel/pkg-fetch). It is used to create and download patched nodejs executables based on the target os/arch and node version.
- creates a snapshot filesystem
- patches `fs` module in order to serve file from the snapshot file system
- creates the final executable

In order to handle native addons, it simply copy them into a temporary folder on startup and the patched fs will return the temporary path when them are required in code.

I think this already covers some limitation of SEA as actually it's missing all the things related to the snapshot file system and it's actually working with a single file. Anyway the hardest part of pkg is to maintain the patched nodejs executables, that's the reason why the project is almost unmaintained right now and all maintainers give up at some point.

If we could find a way to use the features of SEA in pkg without the need of creating custom patched nodejs executable it would be awesome!

**Comment by GabenGar** [comment]
In order to avoid hacking and patching NodeJS runtime you have to implement [Virtual File System](https://github.com/nodejs/single-executable/blob/main/docs/virtual-file-system-requirements.md) first. I don't think there is a support for it by NodeJS as of now, but only because there is no reference implementation either.

**Reply by cspotcode** [reply]
I bet you could use yarn's virtual file system implementation.  I haven't looked at this stuff in forever, but these tickets might have some useful goodies: https://github.com/nexe/nexe/pull/737 https://github.com/nexe/nexe/pull/741

**Reply by arcanis** [reply]
Yeah, I made a sea implementation with it a while ago: https://github.com/arcanis/node-sea

**Reply by robertsLando** [reply]
@arcanis what are the main limitations of node-sea? Apart from the ones listed in readme? Any plan on adding native addons suppport too?

**Reply by guymguym** [reply]
Hey @robertsLando @arcanis very interesting! 

We are using pgk in [noobaa-core](https://github.com/noobaa/noobaa-core/blob/cdd266f64904a67b985fd07c9d97abfe85f43790/package.json#L21) and understand the difficulty to maintain the patched nodejs versions, so were very happy to see SEA starting to become an option.

Worth mentioning the approach of https://github.com/chjj/bpkg that included the native addons in base64 and loaded on demand...

Thanks!

**Reply by roberto-slopez** [reply]
Are you currently using bpkg in your projects? I was able to compile a complex project, but I am unable to load the native modules. Everything else is working fine.

**Reply by guymguym** [reply]
@roberto-slopez no, we haven't actually tried bpgk, I just thought the approach is interesting as it embeds native addons as if it was statically compiled into the executable.

**Reply by JoCat** [reply]
@roberto-slopez I've been researching this topic (using SEA to create an executable) all day today and I finally managed to connect an external module (in this case webview) only thanks to [bpkg](https://github.com/chjj/bpkg)

**Reply by roberto-slopez** [reply]
@JoCat Oh, that sounds great, could you give me an example of how you did it?

**Reply by JoCat** [reply]
@roberto-slopez Yes, I'll have an example repository ready in half an hour.

**Reply by JoCat** [reply]
@roberto-slopez https://github.com/AuroraTeam/Glacier/tree/dev
Run `npm i` and `npm run test`, this will create the application

**Reply by boblund** [reply]
I tried building this on a Mac M1 and got Error: Cannot find module '@glacier-app/webview-darwin-arm64'. Any suggestions?

**Reply by JoCat** [reply]
> I tried building this on a Mac M1 and got Error: Cannot find module '@glacier-app/webview-darwin-arm64'. Any suggestions?

Oh, I didn't compile the files for arm64, I'll take a look at it in a few days

**Reply by JoCat** [reply]
> I tried building this on a Mac M1 and got Error: Cannot find module '@glacier-app/webview-darwin-arm64'. Any suggestions?

You can update the library to version 0.4.0, in which I added support for Mac M1

**Reply by robertsLando** [reply]
I have added a `--sea` option to pkg that allows to build executables using nodejs sea but that is still much limited compared to pkg, support for VFS is still missing BTW

https://github.com/yao-pkg/pkg

**Reply by JoCat** [reply]
> I have added a `--sea` option to pkg that allows to build executables using nodejs sea but that is still much limited compared to pkg, support for VFS is still missing BTW

Yes, I saw and checked it, but I haven't started using this mode yet because there is a problem with packing .node files. I will try to solve this problem in the near future

**Reply by robertsLando** [reply]
@JoCat simply use bpkg before using sea

**Reply by JoCat** [reply]
> @JoCat simply use bpkg before using sea

Yes, that's how it used to be, but I changed something in the process, and this stage of the pipeline disappeared for some unknown reason. I'll figure it out later
